Sources/NIOSOCKS/State/ClientStateMachine.swift

```diff
@@ -0,0 +1,150 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+public enum ClientState: Hashable {
+    case waitingForClientGreeting
+    case waitingForAuthenticationMethod(ClientGreeting)
+    case pendingAuthentication
+    case waitingForClientRequest
+    case waitingForServerResponse(ClientRequest)
+    case active
+}
+
+public struct ConnectionStateError: Error, Hashable {
+    var expected: ClientState
+    var actual: ClientState
+    
+    public init(expected: ClientState, actual: ClientState) {
+        self.expected = expected
+        self.actual = actual
+    }
+}
+
+enum ClientAction: Hashable {
+    case sendGreeting
+    case authenticateIfNeeded(AuthenticationMethod)
+    case sendRequest
+    case proxyEstablished
+    case waitForMoreData
+}
+
+struct ClientStateMachine {
+
+    private var state: ClientState
+    
+    var proxyEstablished: Bool {
+        switch self.state {
+        case .active:
+            return true
+        case .waitingForClientGreeting, .waitingForAuthenticationMethod, .waitingForClientRequest, .waitingForServerResponse, .pendingAuthentication:
+            return false
+        }
+    }
+    
+    var shouldBeginHandshake: Bool  {
+        switch self.state {
+        case .waitingForClientGreeting:
+            return true
+        case .active, .waitingForAuthenticationMethod, .waitingForClientRequest, .waitingForServerResponse, .pendingAuthentication:
+            return false
+        }
+    }
+    
+    init() {
+        self.state = .waitingForClientGreeting
+    }
+    
+}
+
+// MARK: - Incoming
+extension ClientStateMachine {
+    
+    mutating func receiveBuffer(_ buffer: inout ByteBuffer) throws -> ClientAction {
+        switch self.state {
+        case .waitingForAuthenticationMethod(let greeting):
+            return try self.handleSelectedAuthenticationMethod(&buffer, greeting: greeting)
+        case .waitingForServerResponse(let request):
+            return try self.handleServerResponse(&buffer, request: request)
+        default:
+            preconditionFailure("Invalid state")
```

Comment:
I don't think this should crash if we're in the wrong state, this is network-reachable.

Reply:
This is an internal failure though - it means we've set the state wrong somewhere else, and it's not recoverable.

Reply:
It's recoverable without taking the process out: we can kill the connection. I agree that we're in the wrong state, but once we're in the wrong state additional bytes from the network must not crash our process. We either detect the invalid state transition when the user triggers it, or we have to tolerate being in it.